### PR TITLE
Added option to delete original comic when converting comics [184]

### PR DESCRIPTION
**Pull request by BRUCELLA2**
## Status
**READY**

## Migrations
NO

## Description
When converting comics, the admin can choose to delete the original comic

### CODE REVIEW

File: comixed-frontend/src/app/library/actions/library.actions.ts [hunk 1]
```diff
@@ -151,6 +151,7 @@ export class LibraryConvertComics implements Action {
       archiveType: string;
       renamePages: boolean;
       deletePages: boolean;
+      deleteOriginalComic: boolean;
```
Comment (mcpierce) [inline comixed-frontend/src/app/library/actions/library.actions.ts]: Could we rename this parameter to something shorter, like "deleteOriginal"?

File: comixed-frontend/src/app/library/components/convert-comics-settings/convert-comics-settings.component.html [hunk 2]
```diff
@@ -24,6 +24,14 @@
                   [binary]='true'
                   formControlName='deletePages'></p-checkbox>
     </div>
+    <div class='ui-g-12'>
+      <label for='delete-pages'>Delete comic</label>
```
Comment (mcpierce) [inline comixed-frontend/src/app/library/components/convert-comics-settings/convert-comics-settings.component.html]: This text should go into the library translation file.

File: comixed-frontend/src/app/library/components/convert-comics-settings/convert-comics-settings.component.ts [hunk 3]
```diff
@@ -45,7 +45,8 @@ export class ConvertComicsSettingsComponent implements OnInit, OnDestroy {
     this.conversionForm = this.formBuilder.group({
       archiveType: ['', Validators.required],
       renamePages: ['', Validators.required],
-      deletePages: ['', Validators.required]
+      deletePages: ['', Validators.required],
+      deleteOriginalComic: ['', Validators.required]
```
Comment (mcpierce) [inline comixed-frontend/src/app/library/components/convert-comics-settings/convert-comics-settings.component.ts]: Given that it's a binary checkbox, does it affect the form to make this a required field?

File: comixed-model/src/main/java/org/comixedproject/model/comic/Comic.java [hunk 4]
```diff
@@ -337,7 +337,9 @@
   @JsonView({View.ComicList.class, View.LibraryUpdate.class})
   private Integer duplicateCount;
 
-  @ManyToMany(mappedBy = "comics", cascade = CascadeType.ALL)
+  @ManyToMany(
+      mappedBy = "comics",
+      cascade = {CascadeType.PERSIST, CascadeType.MERGE})
```
Comment (mcpierce) [inline comixed-model/src/main/java/org/comixedproject/model/comic/Comic.java]: Is there a technical benefit to limiting the cascade to only PERSIST and MERGE, implicitly excluding DETACH, REMOVE, etc?

File: comixed-rest-api/src/main/java/org/comixedproject/controller/library/LibraryController.java [hunk 5]
```diff
@@ -133,16 +133,19 @@ public void convertComics(@RequestBody() ConvertComicsRequest request) {
     ArchiveType archiveType = request.getArchiveType();
     boolean renamePages = request.isRenamePages();
     boolean deletePages = request.isDeletePages();
+    boolean deleteOriginalComic = request.isDeleteOriginalComic();
```
Comment (mcpierce) [inline comixed-rest-api/src/main/java/org/comixedproject/controller/library/LibraryController.java]: See previous related comment: can we shorten this to deleteOriginal?

**Review by mcpierce — CHANGES_REQUESTED** [review]
In the commit's first line, please also add a # before the issue number. That way Github will make it a link back to the original ticket.

**Comment by mcpierce** [comment]
@BRUCELLA2 I see the following error in the CI run for this PR:
```
2020-07-31T21:33:30.0348903Z org.springframework.dao.DataIntegrityViolationException: 
2020-07-31T21:33:30.0353123Z could not execute statement; SQL [n/a]; constraint ["FKDKDJNUIKDOG29XSGJEJYNDGBT: PUBLIC.READING_LIST_ENTRIES FOREIGN KEY(COMIC_ID) REFERENCES PUBLIC.COMICS(ID) (1000)"; SQL statement:
2020-07-31T21:33:30.0356617Z delete from comics where id=? [23503-200]]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement
2020-07-31T21:33:30.0358629Z 	at org.comixedproject.repositories.comic.ComicRepositoryTest.testDelete(ComicRepositoryTest.java:573)
2020-07-31T21:33:30.0360952Z Caused by: org.hibernate.exception.ConstraintViolationException: could not execute statement
2020-07-31T21:33:30.0363449Z 	at org.comixedproject.repositories.comic.ComicRepositoryTest.testDelete(ComicRepositoryTest.java:573)
2020-07-31T21:33:30.0369591Z Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: 
2020-07-31T21:33:30.0370179Z Referential integrity constraint violation: "FKDKDJNUIKDOG29XSGJEJYNDGBT: PUBLIC.READING_LIST_ENTRIES FOREIGN KEY(COMIC_ID) REFERENCES PUBLIC.COMICS(ID) (1000)"; SQL statement:
2020-07-31T21:33:30.0371056Z delete from comics where id=? [23503-200]
2020-07-31T21:33:30.0371288Z 	at org.comixedproject.repositories.comic.ComicRepositoryTest.testDelete(ComicRepositoryTest.java:573)
2020-07-31T21:33:30.0371474Z 
2020-07-31T21:33:30.0411630Z [INFO] Running org.comixedproject.repositories.comic.PublisherRepositoryTest
2020-07-31T21:33:30.3621899Z [INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.317 s - in org.comixedproject.repositories.comic.PublisherRepositoryTest
2020-07-31T21:33:30.3622716Z [INFO] Running org.comixedproject.repositories.scraping.ScrapingCacheRepositoryTest
2020-07-31T21:33:30.7332453Z [INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.368 s - in org.comixedproject.repositories.scraping.ScrapingCacheRepositoryTest
2020-07-31T21:33:31.2729420Z [INFO] 
2020-07-31T21:33:31.2732031Z [INFO] Results:
2020-07-31T21:33:31.2761202Z [INFO] 
2020-07-31T21:33:31.2761472Z [ERROR] Errors: 
2020-07-31T21:33:31.2762304Z [ERROR]   ComicRepositoryTest.testDelete:573 » DataIntegrityViolation could not execute ...
2020-07-31T21:33:31.2762498Z [INFO] 
2020-07-31T21:33:31.2762781Z [ERROR] Tests run: 84, Failures: 0, Errors: 1, Skipped: 0
```

File: comixed-frontend/src/app/library/actions/library.actions.ts (hunk 1 above)
Comment (BRUCELLA2) [reply comixed-frontend/src/app/library/actions/library.actions.ts]: ok

File: comixed-frontend/src/app/library/components/convert-comics-settings/convert-comics-settings.component.html (hunk 2 above)
Comment (BRUCELLA2) [reply comixed-frontend/src/app/library/components/convert-comics-settings/convert-comics-settings.component.html]: i forgot to change this.

File: comixed-model/src/main/java/org/comixedproject/model/comic/Comic.java (hunk 4 above)
Comment (BRUCELLA2) [reply comixed-model/src/main/java/org/comixedproject/model/comic/Comic.java]: I have a bug and I try something but it doesn't work (break an unit test). CascadeType.ALL is better.

File: comixed-rest-api/src/main/java/org/comixedproject/controller/library/LibraryController.java (hunk 5 above)
Comment (BRUCELLA2) [reply comixed-rest-api/src/main/java/org/comixedproject/controller/library/LibraryController.java]: ok

File: comixed-frontend/src/app/library/components/convert-comics-settings/convert-comics-settings.component.ts (hunk 3 above)
Comment (BRUCELLA2) [reply comixed-frontend/src/app/library/components/convert-comics-settings/convert-comics-settings.component.ts]: After testing, with binary checkbox, required field is not necessary.

**Comment by BRUCELLA2** [comment]
There is a problem with code coverage : 
[WARNING] Could not resolve 1 file paths in [/home/runner/work/comixed/comixed/comixed-frontend/coverage/lcov.info]
[WARNING] First unresolved path: /home/runner/work/comixed/comixed/comixed-frontend/src/test/testing-utils.ts (Run in DEBUG mode to get full list of unresolved paths)

**Comment by BRUCELLA2** [comment]
And another problem, the tests for the class ConvertComicWorkerTask do not seem to have been taken into account.

**Comment by mcpierce** [comment]
Looks good, did some manual testing locally and everything worked as expected. It does expose one issue we'll need to solve: when a comic is deleted from the library there's no way currently to communicate that to the frontend. So after converting a comic and deleting the original, the user may get confused since the original comic stays in the browser until the user refereshes the page. But we can solve that in another ticket.

**Review by mcpierce — APPROVED** [review]
